### DISCUSSION TITLE
Why is baseUrl: "." not actually needed?

### DISCUSSION THREAD

**Discussion by andrechalella** [vercel/next.js#36718] [Help]
Yesterday I finished Next.js tutorials and today I began writing my first real app. One thing I didn't like during the tutorials was having to write relative imports like:

`import Layout from '../../components/layout';`

So first thing I did today was to look for a way to avoid that, and sure enough I found about `jsconfig.json` and `"baseUrl": "."`.

However... to my surprise, I found it is actually not needed. Absolute import paths actually do work for me out-of-the-box!

Consider the tree:

```
.
├── components
│   └── layout.js
└── pages
    ├── components
    │   └── layout.js
    └── index.js
```

Consider the components:

`components/layout.js` returns `<h1>Hello World!</h1>`
`pages/components/layout.js` returns `<h1>Hello again World!</h1>`

I did the following tests with `index.js`.

1. `import Layout from '../components/layout'` gave `Hello World!` (as expected)
2. `import Layout from './components/layout'` gave `Hello again World!` (as expected)
3. `import Layout from 'components/layout'` gave `Hello World!` (NOT expected)
4. `import Layout from '/components/layout'` gave `Hello World!` (wasn't sure what to expect from this one)
5. `import Layout from 'pages/components/layout'` gave `Hello again World!` (expected this to fail compilation)
6. `import Layout from '/pages/components/layout'` gave `Hello again World!` (wasn't sure again)

Neither test changed when I made a `jsconfig.json` file in the root, whatever its value for `baseURL` was.

Thinking a little, I figured this wasn't so much unexpected, since non-relative imports like `next/head` do work. However, I expected the default behavior to be, at least, searching FIRST in the current directory, THEN in root -- that is, I very much expected Test #3 to be `Hello again World!`.

Can somebody shed some light on this? If this really is the expected behavior, then why are relative imports always used in the tutorials, and why is `baseUrl` considered so important?

PS: my Next.js version is 12.1.6

**Comment by ephraimduncan** [comment]
Hello, Can you post your `next.config.js` and `jsconfig.json` here and if you are using typescript, your `tsconfig.json` here.

**Reply by andrechalella** [reply]
I cannot reproduce the behavior anymore. Now, it seems that `baseUrl` in `jsconfig.json` is needed for the imports to work.

I am considering the possibility that my setup has caching issues and, in the tests above, was stuck with `"baseUrl": "."` even after I modified or deleted `jsconfig.json`. During the tutorials, I had two compilation errors that only went away after several minutes (hours in one case) of my correcting the mistake in the code, even though I restarted npm run dev manually several times.

I am on Windows (WSL) and followed [Microsoft's guide](https://docs.microsoft.com/en-us/windows/dev-environment/javascript/nextjs-on-wsl) to install Next.js.

How could I be certain to not be affected by these caching issues in the future? Is there any way to ensure `npm run dev` reads everything from scratch?

**Reply by icyJoseph** [reply]
Hi, 

Changing `jsconfig.js`, `tsconfig.json`, or `next.config.js` among various other config files, requires a full server restart. 

Sometimes you even need to reload VSCode, but in your case, you are assuming that the config files are watched, but that's not true. A way to think about this is that, even if they are watched, the `dev` command is at a stage where it cannot unwind itself to load the config. So you have to manually restart the command.

**Reply by andrechalella** [reply]
Hi.

Thank you for the reply.

I restarted the server, not once but multiple times, among my tests (Ctrl-C + `npm run dev`). Is this not supposed to be enough?

**Reply by icyJoseph** [reply]
It should be enough. I'd recommend trying again. Also, it is `baseUrl`, not `baseURL`.
